gas-retention-simulator/src/main.jsx

```diff
@@ -114,37 +120,71 @@ class GasRetentionSimulator extends React.Component {
         this.drawGasBars(app);
     }
     componentDidUpdate(prevProps, prevState) {
-        if (prevState.activeGases.length !== this.state.activeGases.length) {
+        if (
+            prevState.activeGases.length !== this.state.activeGases.length ||
+            prevState.selectedActiveGas !== this.state.selectedActiveGas
+        ) {
             this.drawGasBars(this.app);
         }
     }
 
+    drawGasBar(app, gas, x) {
+        const g = new PIXI.Graphics();
+        g.interactive = true;
+        g.cursor = 'pointer';
+        g.name = gas.id;
+        g.lineStyle(1, '0x' + toPaddedHexString(gas.color, 6));
+
+        let gasBarOpacity = 0.5;
+        if (this.state.selectedActiveGas === gas.id) {
+            gasBarOpacity = 1;
+        }
+
+        g.beginFill(gas.color, gasBarOpacity);
+
+        g.drawRect(x, 15, 20, this.gasBarHeight);
+        g.endFill();
+
+        if (this.state.activeGases.length === 1) {
+            g.position.x += 52;
+        } else if (this.state.activeGases.length === 2) {
+            g.position.x += 30;
+        } else {
+            g.position.x += 11;
+        }
+
+        app.stage.addChild(g);
+
+        g.on('click', this.onGasBarClick)
+
+          .on('mousedown', this.onGasBarDragStart)
+          .on('touchstart', this.onGasBarDragStart)
+
+          .on('mouseup', this.onGasBarDragEnd)
+          .on('mouseupoutside', this.onGasBarDragEnd)
+
+          .on('touchend', this.onGasBarDragEnd)
+          .on('touchendoutside', this.onGasBarDragEnd)
+
+          .on('mousemove', this.onGasBarMove)
+          .on('touchmove', this.onGasBarMove);
+    }
+
     drawGasBars(app) {
         clearStage(app);
 
-        let g, myGas;
+        let myGas;
         let i = 0;
 
         for (myGas in this.state.activeGases) {
             let gas = this.state.activeGases[myGas];
-            g = new PIXI.Graphics();
-            g.interactive = true;
-            g.cursor = 'pointer';
-            g.lineStyle(1, '0x' + toPaddedHexString(gas.color, 6));
-            g.beginFill(gas.color, 0.5);
-            g.drawRect(40 * i, 2, 20, 130);
-            g.endFill();
-
-            if (this.state.activeGases.length === 1) {
-                g.position.x += 50;
-            } else if (this.state.activeGases.length === 2) {
-                g.position.x += 20;
-            }
 
-            app.stage.addChild(g);
+            this.drawGasBar(app, gas, i * 40);
```

Comment:
`drawGasBars` and `drawGasBar` don't bind `this` in the constructor, is that ok?

Reply:
That's just how you refer to methods defined off the class here.. you could define them in the constructor, or ES6-style like I've done here.

```
class GasRetentionSimulator extends React.Component {
    drawGasBar() {
        // etc
    }
    drawGasBars() {
        // etc
    }
}
```

Reply:
So is the classes `this` in scope for class methods that don't explicitly bind `this`? I guess I'm trying to understand for myself when I need to bind `this` and when I don't.

Reply:
So yeah, looking at this example it looks like you don't: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Prototype_methods

Reply:
If you're talking about this kind of thing I'm doing in the constructor:
```
         this.onAddGas = this.onAddGas.bind(this);
         this.onRemoveGas = this.onRemoveGas.bind(this);
         this.onGasClick = this.onGasClick.bind(this);

         this.onGasBarClick = this.onGasBarClick.bind(this);
         this.onGasBarDragStart = this.onGasBarDragStart.bind(this);
         this.onGasBarDragEnd = this.onGasBarDragEnd.bind(this);
         this.onGasBarMove = this.onGasBarMove.bind(this);
```

That's necessary when binding event callbacks that need to refer to `this` in their function body. `drawGasBar()` is just a normal class method, so `this` inside this method will always refer to the instance of that class.

I don't fully understand why it's necessary to bind to this for event callbacks, that's just kind of a rule of thumb that I follow.

Reply:
Yeah, thats my understanding too. So when you call `this.drawGasBar` here, the enclosing function doesn't bind `this`.